### ISSUE TITLE
DNS-SD API not exposing multiple IP addresses from node/registry

### ISSUE THREAD

**Issue by rhastie** [sony/nmos-cpp#99]
I'm seeing an issue whereby the DNS-SD API (specifically mDNS to HTTP bridging) is not exposing all of the IP addresses from connected nodes... There is only a single IP address being exposed which contradicts mDNS.

This might be a simple configuration issue on my part... I have just re-done mDNS inside the container so a mis-configuration here may be the culprit. Additionally, I don't believe I need anything special in config.json for the DNS-SD API when starting the registry. Could it be over-writing the entries rather than combining them?

As an example, I have the following node data exposed under mDNS/Avahi. I have multiple IP addresses exposed on both the Sony registry and the Sony node.

Ran from the host container that's running the registry:
```
root@leaf:/home# avahi-browse -r _nmos-node._tcp
+ swid0_eth.5 IPv4 nmos-cpp_node_192-168-6-12:11000              _nmos-node._tcp                                                                        local
+ swid0_eth.5 IPv4 nmos-cpp_node_192-168-6-12:8010               _nmos-node._tcp                                                                        local
+ swid0_eth.1 IPv4 nmos-cpp_node_192-168-6-12:11000              _nmos-node._tcp                                                                        local
+ swid0_eth.1 IPv4 emsfp-a0-21-92                                _nmos-node._tcp                                                                        local
+ swid0_eth.1 IPv4 nmos-cpp_node_192-168-6-12:8010               _nmos-node._tcp                                                                        local
+  mgmt0 IPv6 nmos-cpp_node_192-168-6-12:11000              _nmos-node._tcp                                                                        local
+  mgmt0 IPv6 nmos-cpp_node_192-168-6-12:8010               _nmos-node._tcp                                                                        local
+  mgmt0 IPv4 nmos-cpp_node_192-168-6-12:11000              _nmos-node._tcp                                                                        local
+  mgmt0 IPv4 nmos-cpp_node_192-168-6-12:8010               _nmos-node._tcp                                                                        local
+     lo IPv4 nmos-cpp_node_192-168-6-12:11000              _nmos-node._tcp                                                                        local
+     lo IPv4 nmos-cpp_node_192-168-6-12:8010               _nmos-node._tcp                                                                        local
= swid0_eth.1 IPv4 emsfp-a0-21-92                                _nmos-node._tcp                                                                        local
   hostname = [emsfp-a0-21-92.local]
   address = [192.168.39.11]
   port = [80]
   txt = ["api_ver=v1.2" "api_proto=http"]
=  mgmt0 IPv6 nmos-cpp_node_192-168-6-12:11000              _nmos-node._tcp                                                                        local
   hostname = [leaf.local]
   address = [fe80::526b:4bff:fe16:6294]
   port = [11000]
   txt = ["api_auth=false" "api_ver=v1.0,v1.1,v1.2,v1.3" "api_proto=http"]
=  mgmt0 IPv6 nmos-cpp_node_192-168-6-12:8010               _nmos-node._tcp                                                                        local
   hostname = [leaf.local]
   address = [fe80::526b:4bff:fe16:6294]
   port = [8010]
   txt = ["api_auth=false" "api_ver=v1.0,v1.1,v1.2,v1.3" "api_proto=http"]
= swid0_eth.5 IPv4 nmos-cpp_node_192-168-6-12:11000              _nmos-node._tcp                                                                        local
   hostname = [leaf.local]
   address = [192.168.0.1]
   port = [11000]
   txt = ["api_auth=false" "api_ver=v1.0,v1.1,v1.2,v1.3" "api_proto=http"]
= swid0_eth.5 IPv4 nmos-cpp_node_192-168-6-12:8010               _nmos-node._tcp                                                                        local
   hostname = [leaf.local]
   address = [192.168.0.1]
   port = [8010]
   txt = ["api_auth=false" "api_ver=v1.0,v1.1,v1.2,v1.3" "api_proto=http"]
= swid0_eth.1 IPv4 nmos-cpp_node_192-168-6-12:11000              _nmos-node._tcp                                                                        local
   hostname = [leaf.local]
   address = [192.168.39.1]
   port = [11000]
   txt = ["api_auth=false" "api_ver=v1.0,v1.1,v1.2,v1.3" "api_proto=http"]
= swid0_eth.1 IPv4 nmos-cpp_node_192-168-6-12:8010               _nmos-node._tcp                                                                        local
   hostname = [leaf.local]
   address = [192.168.39.1]
   port = [8010]
   txt = ["api_auth=false" "api_ver=v1.0,v1.1,v1.2,v1.3" "api_proto=http"]
=  mgmt0 IPv4 nmos-cpp_node_192-168-6-12:11000              _nmos-node._tcp                                                                        local
   hostname = [leaf.local]
   address = [192.168.6.12]
   port = [11000]
   txt = ["api_auth=false" "api_ver=v1.0,v1.1,v1.2,v1.3" "api_proto=http"]
=  mgmt0 IPv4 nmos-cpp_node_192-168-6-12:8010               _nmos-node._tcp                                                                        local
   hostname = [leaf.local]
   address = [192.168.6.12]
   port = [8010]
   txt = ["api_auth=false" "api_ver=v1.0,v1.1,v1.2,v1.3" "api_proto=http"]
=     lo IPv4 nmos-cpp_node_192-168-6-12:11000              _nmos-node._tcp                                                                        local
   hostname = [localhost]
   address = [127.0.0.1]
   port = [11000]
   txt = ["api_auth=false" "api_ver=v1.0,v1.1,v1.2,v1.3" "api_proto=http"]
=     lo IPv4 nmos-cpp_node_192-168-6-12:8010               _nmos-node._tcp                                                                        local
   hostname = [localhost]
   address = [127.0.0.1]
   port = [8010]
   txt = ["api_auth=false" "api_ver=v1.0,v1.1,v1.2,v1.3" "api_proto=http"]
```

And this is what's getting exposed on http://192.168.6.12:8010/x-dns-sd/v1.0/_nmos-node._tcp/

```
[
{
"addresses": [
"192.168.39.11"
],
"host_target": "emsfp-a0-21-92.local.",
"name": "emsfp-a0-21-92",
"port": 80,
"txt": {
"api_proto": "http",
"api_ver": "v1.2"
}
},
{
"addresses": [
"192.168.0.1"
],
"host_target": "leaf.local.",
"name": "nmos-cpp_node_192-168-6-12:11000",
"port": 11000,
"txt": {
"api_auth": "false",
"api_proto": "http",
"api_ver": "v1.0,v1.1,v1.2,v1.3"
}
},
{
"addresses": [
"192.168.0.1"
],
"host_target": "leaf.local.",
"name": "nmos-cpp_node_192-168-6-12:8010",
"port": 8010,
"txt": {
"api_auth": "false",
"api_proto": "http",
"api_ver": "v1.0,v1.1,v1.2,v1.3"
}
}
]
```

Data is correct on the registry Query API interface - http://192.168.6.12:8010/x-nmos/query/v1.3/nodes/

```
[
{
"api": {
"endpoints": [
{
"host": "192.168.6.12",
"port": 11000,
"protocol": "http"
},
{
"host": "192.168.39.1",
"port": 11000,
"protocol": "http"
},
{
"host": "192.168.0.1",
"port": 11000,
"protocol": "http"
}
],
"versions": [
"v1.0",
"v1.1",
"v1.2",
"v1.3"
]
},
"caps": {},
"clocks": [
{
"name": "clk0",
"ref_type": "internal"
}
],
"description": "ubuntu-Node",
"hostname": "leaf.local",
"href": "http://192.168.6.12:11000/",
"id": "d288f18d-8091-54fe-96e3-b713dac24c60",
"interfaces": [
{
"chassis_id": "192.168.6.12",
"name": "mgmt0",
"port_id": "50-6b-4b-16-62-94"
},
{
"chassis_id": "192.168.39.1",
"name": "swid0_eth.1",
"port_id": "ec-0d-9a-fc-d0-c8"
},
{
"chassis_id": "192.168.0.1",
"name": "swid0_eth.5",
"port_id": "ec-0d-9a-fc-d0-c8"
}
],
"label": "ubuntu-Node",
"services": [],
"tags": {},
"version": "1573853657:974743250"
},
{
"api": {
"endpoints": [
{
"host": "192.168.6.12",
"port": 8010,
"protocol": "http"
},
{
"host": "192.168.39.1",
"port": 8010,
"protocol": "http"
},
{
"host": "192.168.0.1",
"port": 8010,
"protocol": "http"
}
],
"versions": [
"v1.0",
"v1.1",
"v1.2",
"v1.3"
]
},
"caps": {},
"clocks": [],
"description": "leaf",
"hostname": "leaf.local",
"href": "http://192.168.6.12:8010/",
"id": "ceab4123-bd95-59ac-a3c1-76f80792530e",
"interfaces": [
{
"chassis_id": "192.168.6.12",
"name": "mgmt0",
"port_id": "50-6b-4b-16-62-94"
},
{
"chassis_id": "192.168.39.1",
"name": "swid0_eth.1",
"port_id": "ec-0d-9a-fc-d0-c8"
},
{
"chassis_id": "192.168.0.1",
"name": "swid0_eth.5",
"port_id": "ec-0d-9a-fc-d0-c8"
}
],
"label": "leaf",
"services": [
{
"href": "http://192.168.6.12:8010/x-dns-sd/v1.0",
"type": "urn:x-dns-sd/v1.0"
},
{
"href": "http://192.168.39.1:8010/x-dns-sd/v1.0",
"type": "urn:x-dns-sd/v1.0"
},
{
"href": "http://192.168.0.1:8010/x-dns-sd/v1.0",
"type": "urn:x-dns-sd/v1.0"
}
],
"tags": {},
"version": "1573853467:590951449"
}
]
```

**Comment by garethsb-sony** [comment]
Thanks for the report, @rhastie. @rufusu-sony has reproduced this.

By running a few concurrent `avahi-publish` instances to advertise a service and multiple IP addresses for a hostname, we can remove NMOS code from the equation completely.

Although `avahi-browse -r` returns all addresses, we found `avahi-resolve-host-name` only returns a single address, just as does `getent hosts` and library functions like `getaddrinfo` or `gethostbyname` all of which go via the NSS.

This seems to be a known "won't fix" Avahi issue - it intentionally only returns one interface via the higher-level interface of the Name Service Switch, and which one is not guaranteed. See e.g. this thread, [avahi - Resolve multiple IP4 Addresses](https://lists.freedesktop.org/archives/avahi/2009-May/001659.html), from 2009.

This is counter to the Bonjour/mDNSResponder behaviour which provides `DNSServiceGetAddrInfo` and reports all the addresses via this API, but Avahi's compatibility layer doesn't implement it (which is why we use `getaddrinfo`, as recommended by Avahi, instead). The mailing list thread linked above goes some way to explain their recalcitrance.

We can confirm multiple IP addresses _are_ being advertised for the `<hostname>.local` address with either `dig <hostname>.local @224.0.0.251 -p 5353` or `nslookup -port=5353 <hostname>.local 224.0.0.251`.

----

Ideally, I suspect, the DNS-SD API would return services and addresses that would be accessible to the host making the API request; returning **all** addresses was a simple solution for this, which worked with Bonjour/mDNSResponder, but doesn't with Avahi. 😞

One possibility would be to add a new DNS-SD implementation using the lower-level `avahi_service_*` APIs, along side the dns_sd.h API implementation.

However, as far as I can tell this issue will only affect the IP addresses identified for the hostname of the container itself, and only in mDNS mode, so the priority of this issue isn't very high.

**Comment by garethsb-sony** [comment]
We've all done some more thorough testing with multi-homed instances, and found there were some issues even with a fully patched mDNSResponder-based DNS-SD API. These have now largely been solved, by commit 6a26bdbb6d8ea636b4f81672663384b063bca2ff. @rhastie has seen some inconsistent behaviour when there are multiple IP addresses assigned to the same interface, which I think may be related to zero timeouts on the DNSServiceGetAddrInfo API calls. @rufusu-sony has also found that in some circumstances the overall time taken to respond to the DNS-SD API request in his test environment seems to have gone up. I'm going to close this issue, but we may have some follow up ones...
